Adding 'Accept' header to HTTP Probe

fonsecas72 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
When using a HTTP probe, the request will have a "Accept" header by default with the `"*/*"` (accept all)
Most tools do add this header (see curl) so from my perspective it's a reasonable expectation that http probe has it as well.

**Which issue(s) this PR fixes**:
Fixes #95596

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Changed: default "Accept: */*" header added to HTTP probes. See https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#http-probes (https://github.com/kubernetes/website/pull/24756)
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

fonsecas72 commented:
/assign @vishh

fonsecas72 commented:
ping @feiskyer @Random-Liu , can you check this and make it run the tests for me please ?

dims commented:
/ok-to-test
/assign @sjenning @tallclair

dims commented:
/unassign @vishh

dims commented:
/sig node
/sig network

fonsecas72 commented:
/retest

SergeyKanzhelev commented on pkg/probe/http/http.go:
```diff
@@ -101,6 +101,14 @@ func DoHTTPProbe(url *url.URL, headers http.Header, client GetHTTPInterface) (pr
+		// Accept header was not defined. accept all
+		headers.Set("Accept", "*/*")
+	}
+	if headers.Get("Accept") == "" {
```
this is just a theoretical possibility, correct?

SergeyKanzhelev replied:
oh, taking it back. Why do you think we need to override the explicit setting of Accept header to the empty value?

SergeyKanzhelev replied:
is it intended as a way to explicitly delete the header from the yaml?

fonsecas72 replied:
Yes, my idea was to apply the same logic as with the User-Agent and allow the user to explicitly remove the header if he wants to

SergeyKanzhelev commented:
> Does this PR introduce a user-facing change?:

I think it IS user-facing change. It is likely no-op for most, but it will be a change for anybody analyzing headers of probes

fonsecas72 commented:
Thanks!

> I think it IS user-facing change

I do tend to agree. Will add a release note and will document the default headers properly as this is missing.

> Maybe somebody who analyzed bottlenecks can comment on this

@SergeyKanzhelev what should I do about it ? Should I ping someone to take a look? 

IMO it should not be an issue and users can always choose to remove the header.

SergeyKanzhelev commented:
> @SergeyKanzhelev what should I do about it ? Should I ping someone to take a look?
> 
> IMO it should not be an issue and users can always choose to remove the header.

I will ask at tomorrow's SIG node meeting

fonsecas72 commented:
Hello @SergeyKanzhelev, shall we move forward ? :) I have started with the documentation part. 

I'll look into that further if you think this PR is looking good for next steps

SergeyKanzhelev commented on pkg/probe/http/http_test.go:
```diff
@@ -165,6 +165,55 @@ func TestHTTPProbeChecker(t *testing.T) {
 			accBody:    "User-Agent: kube-probe/",
 		},
+		{
+			handler: headerEchoHandler,
```
from test coverage perspective I'm surprised there is no test that validates that there are no "unexpected" headers in the payload. All tests validates that adding or removing specific header works.

I realize it's not relevant to this PR, but maybe you be ok adding this validation?

fonsecas72 replied:
What would you consider an invalid header ? 🤔 Isn't any string a valid header ? 

Or do you mean wrong types - not string ?

SergeyKanzhelev replied:
sorry, I meant that I do not see a test that validates that resulting body doesn't have any additional headers. I would expect that test to fail when you would add `Accept` header by default.

SergeyKanzhelev commented on pkg/probe/http/http_test.go:
```diff
@@ -165,6 +165,55 @@ func TestHTTPProbeChecker(t *testing.T) {
+		{
+			handler: headerEchoHandler,
+			reqHeaders: http.Header{
+				"User-Agent": {"foo/1.0"},
```
does this test case has significance comparing to one without User-Agent below?

fonsecas72 replied:
The idea is that since the `headers` object is being updated when there is a custom `Accept` and when there is custom `User-Agent`  I wanted to make sure one didn't ruin the other when both are customized. It could be that for some reason setting a custom `Accept` would delete/change the `User-Agent` and vice-versa. 

Another option would (/can) be to improve the assert function to allow assertion of 2 headers in same test but thought it would be easier like this.

SergeyKanzhelev replied:
perhaps then add another test with non-specific header `Foo` and `Accept` together. Not critical

fonsecas72 replied:
got it, will look into both suggestions. thanks!

fonsecas72 replied:
Ok, I added tests to verify the default headers and a test with a custom header + Accept

Please take a look

SergeyKanzhelev replied:
thank you!

SergeyKanzhelev commented:
There were no reaction from SIG node meeting. I am following up internally if there were any concerns with the probes payload. But fairly speaking do not expect this would be a blocker.

For the relnotes - maybe just point to the probes page instead of TODO? Less follow ups will be needed.

I added a comment for the test coverage, but beyond this PR looks good.

fonsecas72 commented:
> For the relnotes - maybe just point to the probes page instead of TODO? Less follow ups will be needed.

Sorry, I didn't understand this. You mean to edit the PR body and link the rel.notes to https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes ?

SergeyKanzhelev commented on pkg/probe/http/http.go:
```diff
@@ -101,6 +101,14 @@ func DoHTTPProbe(url *url.URL, headers http.Header, client GetHTTPInterface) (pr
+	}
 	req.Header = headers
 	if headers.Get("Host") != "" {
 		req.Host = headers.Get("Host")
```
unrelated, I'm surprised that Host is not being removed from the collection here.

fonsecas72 replied:
The Host header is required for HTTP 1.1 from what I read. So this should be OK

SergeyKanzhelev replied:
setting `req.Host` must be enough =). It looks like `Host` header in the headers collection will be ignored (https://github.com/golang/go/blob/cad6d1fef5147d31e94ee83934c8609d3ad150b7/src/net/http/request.go#L92). I wasn't sure about whether the header in the collection will be ignored, thus the comment. From code understanding, removing it from the collection would be logical for me here.

SergeyKanzhelev replied:
again, totally irrelevant to this PR.

fonsecas72 replied:
Got it, you mean the condition can be removed because the Headers will always have a Host. Yeah, if you prefer I can remove the condition here or let it for a new PR.

SergeyKanzhelev replied:
no, let's not touch this to minimize changes

SergeyKanzhelev commented:
> > For the relnotes - maybe just point to the probes page instead of TODO? Less follow ups will be needed.
> 
> Sorry, I didn't understand this. You mean to edit the PR body and link the rel.notes to https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes ?

Yes, I meant to include link to the future document: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#http-probes

fonsecas72 committed:
HTTP Probe: Add 'Accept' header by default

When using a HTTP probe, the request will now have a "Accept" header by default with the "*/*" (accept all)
Most tools do add this header (see curl) so it's a reasonable expectation that http probe has it as well

SergeyKanzhelev commented on pkg/probe/http/http_test.go:
```diff
@@ -135,6 +154,18 @@ func TestHTTPProbeChecker(t *testing.T) {
+			handler:    headerKeysNamesHandler,
+			reqHeaders: http.Header{},
+			health:     probe.Success,
+			accBody:    "Accept\nAccept-Encoding\nConnection\nUser-Agent",
```
Interesting, that it already sends `Accept-Encoding` out of the box. I wonder if this is something we need to update in documentation.

Alternative is to remove it when not set explicitly =) Will definitely eliminate any concerns with the size of payload

fonsecas72 replied:
I do tend to agree with the header removal.

Just to be coherent with the previous sentence "let's not touch this to minimize changes" I'll leave it for now and just update the docs.

But anyway I'll be happy to open a new PR to remove this header (+ and the unneeded conditions) or do it in this PR :)

SergeyKanzhelev replied:
let's do documentation first and then follow up on `Accept-Encoding`. It would be great if you can drive it

fonsecas72 replied:
The PR for the docs is including the `Accept-Encoding` header already 
https://github.com/kubernetes/website/pull/24756
 
Yes ofc. I will work on it (removing the `Accept-Encoding` header) right afterwords

SergeyKanzhelev commented:
I think this feature is small enough to not require the enhancement process. With this assumption, this PR is lgtm. One comment is we either need to update proposed docs update to include `Accept-Encoding` or delete `Accept-Encoding` from the default out of the box payload. I'd vote for the removal in this PR, but not strongly.

/lgtm

tallclair commented:
/approve
